resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh

```diff
@@ -53,14 +53,10 @@ fi
 case "$SPARK_K8S_CMD" in
   driver)
     CMD=(
-      ${JAVA_HOME}/bin/java
-      "${SPARK_JAVA_OPTS[@]}"
-      -cp "$SPARK_CLASSPATH"
-      -Xms$SPARK_DRIVER_MEMORY
-      -Xmx$SPARK_DRIVER_MEMORY
-      -Dspark.driver.bindAddress=$SPARK_DRIVER_BIND_ADDRESS
-      $SPARK_DRIVER_CLASS
-      $SPARK_DRIVER_ARGS
+      "$SPARK_HOME/bin/spark-submit"
+      --conf "spark.driver.bindAddress=$SPARK_DRIVER_BIND_ADDRESS"
+      --deploy-mode client
+      "$@"
```

Comment:
There's an argument that can be made for enforcing what arguments can be passed here. For example, we can instead have the following command:

```
      "$SPARK_HOME/bin/spark-submit"
      --conf "spark.driver.bindAddress=$SPARK_DRIVER_BIND_ADDRESS"
      --deploy-mode client
      --properties-file $SPARK_DRIVER_PROPERTIES_FILE
      --class $SPARK_DRIVER_MAIN_CLASS
      spark-internal
```

And then instead of passing args to the Kubernetes container command, we pass everything as environment variables. I think this is clearer and makes the contract more obvious, in terms of what this driver container is expecting as input. Thoughts @vanzin @ifilonenko?

Reply:
What's the problem of passing arguments as arguments? I really dislike using environment variables in general, but this would be an especially nasty use for them. And I'm not sure how they make the contract more obvious.

Reply:
If one is going to override the docker image and provide their own implementation, the customizing developer would like to know all of the parameters that they are being passed up-front. Receiving "$@" makes it less obvious that we should be expecting a properties file and the `SparkLauncher.NO_RESOURCE` from spark-submit.

We could do argument parsing here to verify that `--properties-file` etc. are explicitly provided.

Reply:
+1 on explicit envs.

Reply:
> If one is going to override the docker image and provide their own implementation

There is no documentation about that, so how about defining what that contract is supposed to be first? Otherwise we're all just talking about hypotheticals.

My current understanding is that regardless of what's in the image, `entrypoint.sh` defines the entry point's contract, so this is an internal decision between spark-submit code and that shell script.

Reply:
We've seen some users override the entry point to do a bunch of extra bootstrapping before calling into the actual command to run the driver in custom images. I vaguely recall something related to Openshift doing this - @erikerlandson.

The point is, even if `entrypoint.sh` defines the contract, what arguments would a custom implementation pass it? Something like this would fail because we didn't pass the appropriate arguments to the entrypoint:

```
#!/bin/bash
# my-entrypoint.sh
source /opt/my-custom.script.sh
/opt/spark/entrypoint.sh # Notice we don't pass arguments
```

So the contract may be, that whatever arguments the user's entrypoint is passed, they must eventually call `entrypoint.sh` with all of those arguments forwarded through. That might be an acceptable contract. With environment variables we can say that it doesn't matter what arguments the downstream entrypoint is passed, just that you eventually need to run that code and it knows what to do if the application submission code set up the container's environment properly.

Having defined these ideas more precisely, I think either contract is acceptable here, with a slight preference towards having Spark interpret the environment rather than the arguments.

Reply:
> The point is, even if entrypoint.sh defines the contract, what arguments would a custom implementation pass it? 

`entrypoint.sh` doesn't really define the contract, it implements the contract. Basically the entry point has to respect whatever the Spark submit code tells it to do.

So define the contract first. If you want users to be able to write their own custom entry points, then document that contract in a public doc, so that it cannot change in future versions of Spark. That was the main reason (well, one of them) why I asked this whole thing to be marked "experimental" in 2.3 - nothing is defined, but there seems to be a lot of tribal knowledge about what people want to do.

That doesn't really work in the long run.

Reply:
Think that makes sense re: defining specifically what a custom implementation has to do. We can follow up on that separately. Also agree that we haven't been too precise about what a custom implementation would look like. There's simple things like adding or modifying the existing content, but in terms of modifying logic we haven't given that enough thought.

For now I think the code we have here will suffice for the immediate need of removing the init-container, and we can follow up later on which path we want to take in this part of the discussion.

Reply:
as @mccheah mentioned, I included some logic on the current entrypoint.sh to allow Spark to work in cases such as an anonymous uid (another way to look at it is managing Spark's long-standing quirk of failing when it can't find a passwd entry).  Putting it in entrypoint.sh was a good way to make sure this happened regardless of how the actual CMD evolved.  A kind of defensive future-proofing, which is important for reference Dockerfiles.  It also provides execution via `tini` as "pid 1", which is considered good standard practice.

All of this is done in part with the expectation that _most_ users are liable to just want to customize their images by building the reference dockerfiles and using those as base images for their own, without modifying the CMD or entrypoint.sh

That said, I think that in terms of formally documenting a container API, entrypoint.sh may be a red herring. In theory, a user _should_ be able to build their own custom container from the ground up, up to and including a different entrypoint, or default entrypoint, etc.

Part of the reason we went with an externalized CMD (instead of creating one in the backend code) was to allow maximum flexibility in how images were constructed.  The back-end provides certain information to the pod.  The "API" is a catalogue of this information, combined with any behaviors that the user's container _must_ implement.  API doc shouldn't assume the existence of entrypoint.sh